Download data feature for widgets

adrianq opened the pull request:
- [x] Run [CoreTests.js Casper Tests](https://github.com/openworm/org.geppetto.frontend/blob/master/src/main/webapp/js/pages/tests/casperjs/README.md), make sure they all pass, attach screenshot
- [x] Run [UITests.js Casper Tests](https://github.com/openworm/org.geppetto.frontend/blob/master/src/main/webapp/js/pages/tests/casperjs/README.md), make sure they all pass, attach screenshot
- [ ] Add coverage for whatever new functionality, to a JUnit test if it's backend, to a Casper Test if it's frontend

adrianq commented:
This PR aims to change Science in general and improve the quality of life of the human race. In order to do so, it adds:
- Download data for widgets
   - If a widget implements downloadData method an icon will be automatically added to the title bar. 
   - A new method createZipFromRemoteFiles has been added to GeppettoUtility: Create a zip file from a set of remote files loading them in async way.
   - Refactor code to use new FileSave library
- WidgetButtonToolBar component
   - Implements a very simple react component to use inside the widgets as button tool bar component
   - Use this component in BigImageViewer and DicomViewer
- BigImageViewer
   - Remove old button toolbar
   - Integrate with the model
   - Bug fixing and minor improvements
- Carousel
   - Implements download method
   - Integrate with the model
- Dicom Viewer
   - Fully Refactor: Much better performance, implements everything in a more reactish way, remove unused code and clean up
   - Bug fixing
   - Integrate with the model
   - Ignore ami.js in transpilation process
- Google Viewer
   - Integrate with the model
   - Bug fixing
   - Remove Google Key from the properties (still pending how to deal with this.... currently using my personal google key)
- Improve casper documentation
- Add Skeleton for casper test for new sample project. Work in progress. Depends on this [branch](https://github.com/openworm/org.geppetto.core/tree/patientHMSample)

adrianq commented:
- Download data for widgets
   - If a widget implements downloadData method an icon will be automatically added to the title bar. 
   - A new method createZipFromRemoteFiles has been added to GeppettoUtility: Create a zip file from a set of remote files loading them in an async way.
   - Refactor code to use new FileSave library
- WidgetButtonToolBar component
   - Implements a very simple react component to use inside the widgets as button toolbar component
   - Use this component in BigImageViewer and DicomViewer
- BigImageViewer
   - Remove old button toolbar
   - Integrate with the model
   - Bug fixing and minor improvements
- Carousel
   - Implements download method
   - Integrate with the model
- Dicom Viewer
   - Fully Refactor: Much better performance, implements everything in a more reactish way, remove unused code and clean up
   - Bug fixing
   - Integrate with the model
   - Ignore ami.js in transpilation process
- Google Viewer
   - Integrate with the model
   - Bug fixing
   - Remove Google Key from the properties (still pending how to deal with this.... currently using my personal Google key)
- Improve Casper documentation
- Refactor less/css
   - Migrate css files into less
   - Create default theme for Geppetto (primary and secondary color, background colors and background color widget)
   - Allows theme customization (modify less loader)
- Refactor Geppetto properties.
   - Create a GeppettoConfiguration.json file joining extensionConfiguration and pom properties. Properties can still be passed through the command line which will override properties defined in the properties file.
   - Add theme property
   - GeppettoConfiguration is now loaded on run time so no need to write properties in ejs files.
- Remove licenses from some files

@mattearnshaw @Robbie1977 @gidilli @matteo @pgleeson! If this PR is finally accepted as is you may need to update your environment. Instead of defining the properties for the frontend in pom.xml or command line I would recommend you to create your own GeppettoConfiguration.json (there is one by default with the default parameters). This file defines all current Geppetto Properties. This file will be converted into a javascript dictionary which can be accessed from anywhere in the js code.
Also, we have introduced the concept of Geppetto theme. If you want to experiment with it just copy [default colors.less](https://github.com/openworm/org.geppetto.frontend/blob/downloadData/src/main/webapp/extensions/geppetto-default/colors.less) to your Geppetto extension, add the path to GeppettoConfiguration.json theme variable and play around with it.

I have also made some small changes in your extension mainly to use color theme properties. I have uploaded them into a branch. It is not compulsory to merge them into your development branch but I would highly recommend it so we don't use anymore hardcode colours and for future development.

@Robbie1977  -> https://github.com/adrianq/geppetto-vfb/tree/lessRefactor
@pgleeson, @mattearnshaw -> https://github.com/OpenSourceBrain/geppetto-osb/tree/lessRefactor
@tarelli -> https://github.com/MetaCell/geppetto-hm/pull/8 (ready to be merged!)

adrianq commented:
@jrmartin @gidili @tarelli 

UI Tests:
- First I got stuck in this [test](https://github.com/openworm/org.geppetto.frontend/blob/master/src/main/webapp/js/pages/tests/casperjs/UITests.js#L47) which seems to require persistence bundle. I commented it out so I can run the rest of tests. Maybe we can add this to another set of tests?
- After, most of tests pass except these two:
![screenshot from 2017-06-29 01 41 56](https://user-images.githubusercontent.com/6850223/27667547-54e4e794-5c73-11e7-9888-1851b013d683.png)
I was reading that code and seems to be fine. It shows or hide some icons on the experiment table depending on the user permissions. I double check with Live and the icons are completely different. Before start debugging... any idea guys why are icons different? Permissions? What should be the expected behaviour?

For CoreTest:
- I am getting an error (of a popup widget not displaying properly) but I think is because the nwb bundle requirement. I tried to deploy the nwb bundle but after 30 minutes of the mvn dance I surrender.
- I have another two error regarding connections lines and number of rows in the control panel that I couldn't reproduce them. Seems to be working fine, I guess it may have to do with the timeouts.

In general, some errors seem to be a bit unstable. I was getting a different result every now and then.

tarelli commented:
@adrianq Thanks! All tests in UITests require the persistence bundle, the name is highly misleading and I think @gidili has already changed it in a branch. The idea is that CoreTests.js work without persistence and UITests work with the persistence bundle. If you don't have persistence bundle then I would expect pretty much everything to fail ([here](https://github.com/openworm/org.geppetto.frontend/blob/master/src/main/webapp/js/pages/tests/casperjs/UITests.js#L220) we are persisting for instance). Regarding the icons visibility if you don't have the persistence bundle then you cant write on the projects and some icons that should be there are not etc. See other runs for other PRs ([here](https://github.com/openworm/org.geppetto.frontend/pull/626) and [here](https://github.com/openworm/org.geppetto.frontend/pull/630)). In particular @gidili has also had problems with the Popup test...while Matt has not...maybe Matt has the NWB bundle? @jrmartin is the NWB bundle required for that test? The one about connection lines seems new and neither @gidili or @mattearnshaw have encountered it, what is the check that is failing? 
As for the code itself I will review it on Monday :)

mattearnshaw commented:
@adrianq for NWB, something that caught me out was that the nwb sample files are stored using git LFS, so you should check 

* `org.geppetto.core/src/main/resources/projects/NWBSample/313862020.nwb`
* `org.geppetto.model.nwb/src/main/resources/354190011.nwb`
* `org.geppetto.model.nwb/src/main/resources/313862020.nwb`

have been fetched with git LTS (they are 20MB+ files)

I had problems with `a.deleteIcon` also. For the latter I was just bypassing it because it's a problem with the test, it doesn't "click" properly on the project for some reason. I thought this might be because I had to force casper to work with my version of firefox and something was broken, but it seems like the test needs to be changed. Overall it has been quite a headache getting things to pass, with inconsistencies between runs.

tarelli commented:
Let's make a list of the "unstable" tests and let's fix them, this was to be expected since it's the first time everyone is start to run things :) cc @gidili @jrmartin

adrianq commented:
UITests
This is what I get if I comment out the project with id =2 as I don't have the persistence bundle deployed.
![screenshot from 2017-06-29 12 55 47](https://user-images.githubusercontent.com/6850223/27686856-7a5a7562-5ccc-11e7-9618-3196b838d27b.png)

NWB
I saw that some tests where failing due to resources missing and just commented them out. Afterwards, I couldn't make it work but it seems to me it was more about mvn installing a few more times. I would remove this test from the core tests though.

@mattearnshaw I also had to force casper to work with the latest Firefox. I updated the [doc](https://github.com/openworm/org.geppetto.frontend/blob/downloadData/src/main/webapp/js/pages/tests/casperjs/README.md) I just changed the parameter in application.ini and guessed you did the same...

CoreTests
This is the latest output:
![screenshot from 2017-06-29 13 10 56](https://user-images.githubusercontent.com/6850223/27686908-aa8bbbec-5ccc-11e7-9abc-7f22cbcab136.png)

First error is not true. I checked it manually and it is actually displaying three rows. This is one of the unstables errors: depending on the execution I have 1,2 or 3 errors related to wrong number of rows in the control panel. 

Second one (that happens also for @gidili ) has to do with the NBW bundle. It is checked just after loading the NWB project.

@tarelli as you can see now I can not reproduce the one about the lines ("Right amount of connections line"). Yesterday night it was happening all the time. Another unstable test... I guess the laptop was as tired as me :)

mattearnshaw commented:
I think one cause of unreliability is due to the system mouse being captured by the casper window, which can interfere with things like

`//roll over the experiments row`
`this.mouse.move('tr.experimentsTableColumn:nth-child(1)');`

which is a prerequisite for checking certain icons.

I can't see a way to turn that off, so either test should always be left alone or maybe there are more robust methods for focusing

gidili commented:
@adrianq You cannot run UITests at all without persistence, all it does is load a bunch of projects from urls and persists them.

Good to know the second error (popup) on CoreTests is NWB, because it was the only one I was getting and I didn't have that bundle deployed.

jrmartin commented:
@tarelli. Yes, CoreTests need NWB bundle, there's a basic test that checks the NWB sample core project loads.

jrmartin commented:
@adrianq About the failure on the Control Panel rows, do you get that failure for the same test/project (looks like CA1 project)? Or do you see it happening for other tests too?

adrianq commented:
It is usually that one. I think I have another error every now and then expecting 10 rows but cannot reproduce it anymore. The other one I use to have was with the testingConnectionLines but it is not happening anymore either.

adrianq commented on src/main/webapp/GeppettoConfiguration_production.json:
```diff
@@ -0,0 +1,18 @@
+{
```
No really. As json doesn't support comments I added the production 'template' as another file. I was not sure if it could be interesting to provide this template... we can delete it if we prefer...

tarelli commented:
@adrianq Also found some problems with the colours here 
![screen shot 2017-07-06 at 13 57 43](https://user-images.githubusercontent.com/81127/27912018-378a44ce-6253-11e7-92a9-6ffe57dc6f8a.png)

adrianq commented:
@tarelli what is the problem? The border of the icons?

tarelli commented:
@adrianq The buttons are not visible anymore, the text is all "washed" as if it's behind some layer with some opacity, this is on live:
<img width="641" alt="screen shot 2017-07-06 at 14 51 20" src="https://user-images.githubusercontent.com/81127/27914288-9bf58b7e-625a-11e7-8cb4-7060a8c89eff.png">
